From 037feb5424534a3388ffe12dbf15b4886fcc23bb Mon Sep 17 00:00:00 2001
From: Fabien Dessenne <fabien.dessenne@st.com>
Date: Wed, 2 Jul 2025 17:50:32 +0200
Subject: [PATCH] libcamera: x-linux-isp-v6.1.0

Squashed commit of the following:

commit 8b7e1926bca83625dc6ebdc6855beb66b2f24781
Author: Bich Hemon <bich.hemon@st.com>
Date:   Thu Jun 19 17:18:59 2025 +0200

    ipa: dcmipp: fix awb profile detection

    Increase counter for reconfiguration cases to give more time for AWB algo convergence.
    And disable anti-oscillation after reconfiguration

    Signed-off-by: Bich Hemon <bich.hemon@st.com>

commit afd1d689e816acc7fec2b43fbab6d65f7f31a8cd
Author: Fabien Dessenne <fabien.dessenne@st.com>
Date:   Thu Jun 19 14:29:21 2025 +0200

    ipa: use JudgeII as tuning reference file

    Consider the JudgeII light box as the reference (instead of the E27 light
    box).

    Signed-off-by: Fabien Dessenne <fabien.dessenne@st.com>

commit 5dc489f59daa1bfeb343ddbcbe9bb7afdced3a9b
Author: Fabien Dessenne <fabien.dessenne@st.com>
Date:   Thu Jun 19 14:25:31 2025 +0200

    ipa: fix tuning files

    - Static gain (FixedGainR/G/B) must be used with precision factor
      100000000.
    - Typo fix for "Judge".
    - Fix the Light box name.

    Signed-off-by: Fabien Dessenne <fabien.dessenne@st.com>

commit 81b90d65a583af966a39f381b819075f4cd25e51
Author: Fabien Dessenne <fabien.dessenne@st.com>
Date:   Tue Jun 10 17:00:22 2025 +0200

    ipa: dcmipp: improve sensor delay measure

    Exclude the min and max value from the set of sensor delay measures.
    This makes the measurement less dependent on variations in the light
    environment.

    Signed-off-by: Fabien Dessenne <fabien.dessenne@st.com>

commit 3476dbab8f283adf69fb8d3b34feff6a0e87fee0
Author: Fabien Dessenne <fabien.dessenne@st.com>
Date:   Wed May 28 09:58:02 2025 +0200

    ipa: dcmipp: support AWB Custom color temp in tuning file

    Support "Mode" and "CustomColorTemperature" as YAML tuning file
    parameters.
    When "Mode" is set to AwbCustom (7), the algorithm selects only the
    profile identified by CustomColorTemperature.

    Signed-off-by: Fabien Dessenne <fabien.dessenne@st.com>

commit 75a39a83e232e8b6f515412744e8c151a5c63e7b
Author: Fabien Dessenne <fabien.dessenne@st.com>
Date:   Mon May 26 17:26:58 2025 +0200

    ipa: dcmipp: typo fix

    Set the enable parameter once.

    Signed-off-by: Fabien Dessenne <fabien.dessenne@st.com>

commit 478c6c839220d953d9661f4421ee393cb6397832
Author: Fabien Dessenne <fabien.dessenne@st.com>
Date:   Wed May 7 15:33:38 2025 +0200

    gstreamer: fix memory crash when AWB profiles are not defined

    This workaround fixes an issue where the AWB profiles are not
    defined which happens for example when using the 'uncalibrated'
    tuning file.

    Signed-off-by: Fabien Dessenne <fabien.dessenne@st.com>

commit 21e431d1664938a2afcb7739ef2eaf437da233c2
Author: Fabien Dessenne <fabien.dessenne@st.com>
Date:   Mon May 19 09:48:34 2025 +0200

    ipa: add tuning file for vd56g3 sensor

    Tuning file for the STMicroelectronics VD56G3 / VD66GY camera sensor.

    Signed-off-by: Fabien Dessenne <fabien.dessenne@st.com>

commit 1d86f70a003c3008c6ea7e1d8be2954eb7baa377
Author: Fabien Dessenne <fabien.dessenne@st.com>
Date:   Mon May 5 17:22:16 2025 +0200

    libcamera: libipa: camera_sensor: Add ST VD56G3 sensor properties

    Provide the STMicroelectronics VD56G3 / VD66GY camera sensor
    properties and registration with libipa for the gain code helpers.

    Signed-off-by: Fabien Dessenne <fabien.dessenne@st.com>

commit 154c1a26602515c57134cbb1031c4fcfde572482
Author: Fabien Dessenne <fabien.dessenne@st.com>
Date:   Mon May 5 15:21:37 2025 +0200

    ipa: add uncalibrated.yaml tuning file

    This empty tuning file listing all available algorithms (disabled by
    default) is used when no sensor tuning file is available.

    Signed-off-by: Fabien Dessenne <fabien.dessenne@st.com>

commit 5e3bbbde2473a28a1f6da5afad7370a0e72cfc34
Author: Fabien Dessenne <fabien.dessenne@st.com>
Date:   Tue May 20 09:48:21 2025 +0200

    ipa: dcmipp: disable sensor AE if supported

    If the tuning file specifies that the AE algorithm shall be used, then
    disable the sensor AE if supported (eg vd56g3).

    Signed-off-by: Fabien Dessenne <fabien.dessenne@st.com>

commit c061a62cf7a62fc342bb0b15153c2684f9420aa8
Author: Fabien Dessenne <fabien.dessenne@st.com>
Date:   Tue May 6 15:40:35 2025 +0200

    ipa: dcmipp: set default ISP config to 'disable'

    The default ISP config, which is typically used when the 'uncalibrated'
    tuning file is used shall disable all the ISP controls.
    Fix the demosaicing and gamma default values (disable).
    For demosaicing: enable the feature if the "filters" parameters are
    all defined.

    Signed-off-by: Fabien Dessenne <fabien.dessenne@st.com>

commit 75028857cde59a8537b4d6924553054d200baa24
Author: Fabien Dessenne <fabien.dessenne@st.com>
Date:   Tue May 6 14:53:44 2025 +0200

    ipa: dcmipp: do not use AEC if the sensor CameraHelper is not defined

    The libipa sensor CameraHelpers are used to convert sensor gain
    from their 'code' to their dB value.
    If no CameraHelper is defined for the used sensor, then it is not
    possible to set the sensor gain as expected by the AEC algo : in
    that case, prevent any sensor gain update.
    In addition to that, warn the user with the explicit error log
    "No camera sensor helper for <sensor>. AEC will be disabled"

    Signed-off-by: Fabien Dessenne <fabien.dessenne@st.com>

commit 840830addcc1a2d869039a74037f2af3d35d6736
Author: Fabien Dessenne <fabien.dessenne@st.com>
Date:   Mon May 5 15:21:30 2025 +0200

    pipeline: dcmipp: fallback to uncalibrated tuning file

    If the YAML tuning file isn't found, fall back to the 'uncalibrated'
    configuration file.

    Signed-off-by: Fabien Dessenne <fabien.dessenne@st.com>

commit a662d32a5912f095985932e8c81d6b696e2e00e4
Author: Fabien Dessenne <fabien.dessenne@st.com>
Date:   Mon May 26 09:45:00 2025 +0200

    pipeline: dcmipp: limit generated  configuration to sensor size

    Limit the 'default' generated resolutions to the sensor size.

    Signed-off-by: Fabien Dessenne <fabien.dessenne@st.com>

---
 src/gstreamer/gstlibcamerasrc.cpp             |  3 ++
 src/ipa/dcmipp/algorithms/aec.cpp             | 29 +++++++---
 src/ipa/dcmipp/algorithms/awb.cpp             | 45 +++++++++++++---
 src/ipa/dcmipp/algorithms/blc.cpp             |  1 -
 src/ipa/dcmipp/algorithms/demosaicing.cpp     | 10 +++-
 src/ipa/dcmipp/algorithms/gamma.cpp           |  2 +-
 src/ipa/dcmipp/data/imx335.yaml               | 23 ++++----
 .../dcmipp/data/imx335_judge2_light_box.yaml  | 54 -------------------
 .../dcmipp/data/imx335_mini_light_box.yaml    |  6 +--
 src/ipa/dcmipp/data/meson.build               |  2 +
 src/ipa/dcmipp/data/uncalibrated.yaml         | 21 ++++++++
 src/ipa/dcmipp/data/vd56g3.yaml               | 54 +++++++++++++++++++
 src/ipa/dcmipp/dcmipp.cpp                     | 33 ++++++++----
 src/ipa/dcmipp/ipa_context.h                  |  1 +
 src/ipa/libipa/camera_sensor_helper.cpp       | 11 ++++
 src/libcamera/pipeline/dcmipp/dcmipp.cpp      | 14 +++--
 .../sensor/camera_sensor_properties.cpp       |  4 ++
 17 files changed, 210 insertions(+), 103 deletions(-)
 delete mode 100644 src/ipa/dcmipp/data/imx335_judge2_light_box.yaml
 create mode 100644 src/ipa/dcmipp/data/uncalibrated.yaml
 create mode 100644 src/ipa/dcmipp/data/vd56g3.yaml

diff --git a/src/gstreamer/gstlibcamerasrc.cpp b/src/gstreamer/gstlibcamerasrc.cpp
index 164092bc..f6aaeeee 100644
--- a/src/gstreamer/gstlibcamerasrc.cpp
+++ b/src/gstreamer/gstlibcamerasrc.cpp
@@ -444,6 +444,9 @@ GstLibcameraSrcState::requestCompleted(Request *request)
 	}
 	const auto AwbAlgoProfileNames = request->metadata().get(controls::draft::AwbProfileName);
 	if (AwbAlgoProfileNames) {
+		/* The below "".fill()"" prevents a memory crash happening when we set a profile_name to "" */
+		/* Do not ask me why this workaround fixes the issue : I have no idea and I do not want to know */
+		src_->ctrl.awb_algo_profile_names.fill("X");
 		/* The Controls class does not support array of string. So, split the concatenated string */
 		std::string token;
 		std::stringstream ss(*AwbAlgoProfileNames);
diff --git a/src/ipa/dcmipp/algorithms/aec.cpp b/src/ipa/dcmipp/algorithms/aec.cpp
index 546cd672..80ea43f9 100644
--- a/src/ipa/dcmipp/algorithms/aec.cpp
+++ b/src/ipa/dcmipp/algorithms/aec.cpp
@@ -226,12 +226,12 @@ int Aec::init([[maybe_unused]] IPAContext &context, const YamlObject &tuningData
 	params_.algoTarget = kAlgoIdealExposureTarget * pow(2, params_.algoExposureValue);
 
 	/* Check for sensor static config (exposure time in microseconds, gain in dB) */
-	params_.sensorStatic.exposure = tuningData["ExposureTime"].get<int32_t>(10000);
+	params_.sensorStatic.exposure = tuningData["ExposureTime"].get<int32_t>(context.info.sensorExposureMax);
 	if (!isExposureTimeValid(params_.sensorStatic.exposure, context)) {
 		LOG(DcmippAec, Error) << "Invalid Exposure time: " << params_.sensorStatic.exposure;
 		return -EINVAL;
 	}
-	params_.sensorStatic.gain = tuningData["AnalogueGain_dB"].get<double>(1);
+	params_.sensorStatic.gain = tuningData["AnalogueGain_dB"].get<double>(context.info.sensorGainMin);
 	if (!isGainValid(params_.sensorStatic.gain, context)) {
 		LOG(DcmippAec, Error) << "Invalid Gain: " << params_.sensorStatic.gain;
 		return -EINVAL;
@@ -335,6 +335,7 @@ int Aec::configure(IPAContext &context,
 	}
 
 	/* Update context immediately, so IPA can apply the sensor settings at start up */
+	context.info.AECEnable = params_.algoEnable;
 	context.sensor = config_.sensor;
 	config_.pending = true;
 
@@ -462,6 +463,7 @@ void Aec::prepare(IPAContext &context,
 		sensorControls.set(controls::ExposureTime, config_.sensor.exposure);
 
 		/* Update context */
+		context.info.AECEnable = params_.algoEnable;
 		context.sensor = config_.sensor;
 
 		/* Start the invalid stats counter from this update. Consider both ISP and sensor */
@@ -535,15 +537,26 @@ void Aec::processSensorDelayMeasure(const stm32_dcmipp_stat_buf *stats)
 		internal_.invalidStats = 1;
 	} else {
 		/* All the configurations have been tested, finalize the test procedure */
-		int32_t sum = 0;
-		int n_valid = 0;
-		/* Find the average delay among the valid (!= kAecSensorDelayMax) values */
-		for (int i = 0; i < kAecNbTestConfig - 1; i++)
-			if (internal_.measure.delays[i] != kAecSensorDelayMax) {
-				sum += internal_.measure.delays[i];
+		int32_t delay, min, max, sum = 0, n_valid = 0;
+		min = kAecSensorDelayMax;
+		max = 0;
+		/* Find the average delay among the valid (!= kAecSensorDelayMax) values and also exclude the min and max value */
+		for (int i = 0; i < kAecNbTestConfig - 1; i++) {
+			delay = internal_.measure.delays[i];
+			if (delay != kAecSensorDelayMax) {
+				sum += delay;
 				n_valid++;
+				max = (delay > max) ? delay : max;
+				min = (delay < min) ? delay : min;
 			}
+		}
 		if (n_valid) {
+			if (n_valid >= 4) {
+				/* Exclude min and max */
+				sum -= min;
+				sum -= max;
+				n_valid -= 2;
+			}
 			internal_.measure.sensorDelayMeasure = round((float)sum / n_valid);
 			/* Apply the measure internally */
 			params_.sensorDelay = internal_.measure.sensorDelayMeasure;
diff --git a/src/ipa/dcmipp/algorithms/awb.cpp b/src/ipa/dcmipp/algorithms/awb.cpp
index 006ad4b8..3d05a036 100644
--- a/src/ipa/dcmipp/algorithms/awb.cpp
+++ b/src/ipa/dcmipp/algorithms/awb.cpp
@@ -70,7 +70,7 @@ static evision_api_awb_init_profiles_t evision_api_awb_init_profiles = nullptr;
 static evision_api_awb_run_average_t evision_api_awb_run_average = nullptr;
 
 static constexpr uint8_t kIspStatCheckSkipAfterInit = 10;
-static constexpr uint8_t kIspStatCheckSkipAfterCTEstimation = 4;
+static constexpr uint8_t kIspStatCheckSkipAfterCTEstimation = 6;
 static constexpr int32_t kIspStatDeltaMax = 2;
 
 static int32_t stats_history[3][3] = { 0 };
@@ -469,6 +469,27 @@ int Awb::init([[maybe_unused]] IPAContext &context, const YamlObject &tuningData
 		return -EINVAL;
 	}
 
+	/* Check for Custom mode */
+	int32_t awbMode = tuningData["Mode"].get<int32_t>(controls::AwbAuto);
+	int32_t customColorTemp = tuningData["CustomColorTemperature"].get<int32_t>(kColTemperatureMax);
+	if (awbMode == controls::AwbCustom) {
+		/* Search for the corresponding profile */
+		int profId;
+		for (profId = 0; profId < kAwbNbRef; profId++)
+			if (params_.referenceColorTemp[profId] == customColorTemp)
+				break;
+
+		if (profId >= kAwbNbRef) {
+			LOG(DcmippAwb, Error) << "Invalid Custom Colour Temp: " << customColorTemp;
+			return -EINVAL;
+		}
+		/* Disable auto mode and set customColorTemp */
+		params_.algoEnable = false;
+		internal_.customColorTemp = customColorTemp;
+		LOG(DcmippAwb, Debug) << "Using custom mode for AWB profile " << params_.profileName[profId]
+				      << " (" << customColorTemp << ")";
+	}
+
 	/* Check for ISP Gain / ColorConv static config */
 	params_.gainStatic.gainR = tuningData["FixedGainR"].get<int32_t>(0);
 	params_.gainStatic.gainG = tuningData["FixedGainG"].get<int32_t>(0);
@@ -544,13 +565,23 @@ int Awb::init([[maybe_unused]] IPAContext &context, const YamlObject &tuningData
 int Awb::configure([[maybe_unused]] IPAContext &context,
 		   [[maybe_unused]] const IPACameraSensorInfo &configInfo)
 {
+	int profId;
 	/* Set the initial ISP values in the pending config */
 	if (!params_.algoEnable) {
-		/* Apply static ISP Gain / ColorConv config */
-		config_.gain = params_.gainStatic;
-		config_.cconv = params_.cconvStatic;
+		if (internal_.customColorTemp) {
+			/* Get and apply the profile of the Custom Color temperature */
+			for (profId = 0; profId < kAwbNbRef; profId++)
+				if (params_.referenceColorTemp[profId] == internal_.customColorTemp)
+					break;
+			profId = profId >= kAwbNbRef ? 0 : profId;
+			applyProfile(profId);
+		} else {
+			/* Apply static ISP Gain / ColorConv config */
+			config_.gain = params_.gainStatic;
+			config_.cconv = params_.cconvStatic;
+		}
 	} else {
-		int i, profId;
+		int i;
 		/* Apply the first profile with color temp above minimum value before first processing of the algorithm */
 		profId = 0;
 		for (i = 0; i < kAwbNbRef; i++) {
@@ -853,8 +884,8 @@ void Awb::process([[maybe_unused]] IPAContext &context,
 					/* Run algo to estimate gain and color conversion to apply */
 					e_ret = evision_api_awb_run_average(pIspAWBestimator, NULL, 1, meas);
 					if (e_ret == EVISION_RET_SUCCESS) {
-						if (pIspAWBestimator->out_temp != internal_.colorTemp) {
-							if (pIspAWBestimator->out_temp == color_temperature_history[1]) {
+						if ((pIspAWBestimator->out_temp != internal_.colorTemp) || internal_.profileDirty) {
+							if ((pIspAWBestimator->out_temp == color_temperature_history[1]) && !internal_.profileDirty) {
 								skip_stat_check_count = 0; //oscillation detected
 								LOG(DcmippAwb, Debug) << "Oscillation detected";
 							} else {
diff --git a/src/ipa/dcmipp/algorithms/blc.cpp b/src/ipa/dcmipp/algorithms/blc.cpp
index 9b40dd5f..c633d12b 100644
--- a/src/ipa/dcmipp/algorithms/blc.cpp
+++ b/src/ipa/dcmipp/algorithms/blc.cpp
@@ -43,7 +43,6 @@ int BlackLevelCorrection::init([[maybe_unused]] IPAContext &context, const YamlO
 		params_.blackLevel.blcR = level[0];
 		params_.blackLevel.blcG = level[1];
 		params_.blackLevel.blcB = level[2];
-		params_.blackLevel.enable = false;
 		params_.blackLevel.enable = tuningData["Enable"].get<bool>(true);
 	} else {
 		LOG(DcmippBlc, Error) << "Invalid nubmer of black levels";
diff --git a/src/ipa/dcmipp/algorithms/demosaicing.cpp b/src/ipa/dcmipp/algorithms/demosaicing.cpp
index b0c7915a..e86ec86d 100644
--- a/src/ipa/dcmipp/algorithms/demosaicing.cpp
+++ b/src/ipa/dcmipp/algorithms/demosaicing.cpp
@@ -29,10 +29,15 @@ static bool isValid(int32_t filter)
 int Demosaicing::init([[maybe_unused]] IPAContext &context, const YamlObject &tuningData)
 {
 	/* Parse Tuning Data to get algo parameters */
-	params_.demosaicing.enable = tuningData["Enable"].get<bool>(true);
 	/* ISP Demosaicing Filter */
 	std::vector<int32_t> filter = tuningData["Filter"].getList<int32_t>().value_or(std::vector<int32_t>(kNbFilters, kFilterDef));
-	if (filter.size() != kNbFilters) {
+	if (!filter.size()) {
+		params_.demosaicing.peak = kFilterDef;
+		params_.demosaicing.linev = kFilterDef;
+		params_.demosaicing.lineh = kFilterDef;
+		params_.demosaicing.edge = kFilterDef;
+		params_.demosaicing.enable = false;
+	} else if (filter.size() != kNbFilters) {
 		LOG(DcmippDemosaicing, Error) << "Invalid number of filters";
 		return -EINVAL;
 	}
@@ -46,6 +51,7 @@ int Demosaicing::init([[maybe_unused]] IPAContext &context, const YamlObject &tu
 	params_.demosaicing.linev = filter[1];
 	params_.demosaicing.lineh = filter[2];
 	params_.demosaicing.edge = filter[3];
+	params_.demosaicing.enable = tuningData["Enable"].get<bool>(true);
 
 	/* Configure exposed controls */
 	context.dcmippControls[&controls::draft::DemosaicingEnable] = ControlInfo(false, true);
diff --git a/src/ipa/dcmipp/algorithms/gamma.cpp b/src/ipa/dcmipp/algorithms/gamma.cpp
index fd595479..34c013f6 100644
--- a/src/ipa/dcmipp/algorithms/gamma.cpp
+++ b/src/ipa/dcmipp/algorithms/gamma.cpp
@@ -20,7 +20,7 @@ int GammaCorrection::init([[maybe_unused]] IPAContext &context, const YamlObject
 {
 	/* Parse Tuning Data to get algo parameters */
 	/* ISP Gamma Correction config */
-	params_.gamma.enable = tuningData["Enable"].get<bool>(true);
+	params_.gamma.enable = tuningData["Enable"].get<bool>(false);
 
 	/* Configure exposed controls */
 	context.dcmippControls[&controls::draft::GammaCorrectionEnable] = ControlInfo(false, true);
diff --git a/src/ipa/dcmipp/data/imx335.yaml b/src/ipa/dcmipp/data/imx335.yaml
index 3de5a11b..6e780d10 100644
--- a/src/ipa/dcmipp/data/imx335.yaml
+++ b/src/ipa/dcmipp/data/imx335.yaml
@@ -1,12 +1,11 @@
-# Copyright (c) 2024 STMicroelectronics.
+# Copyright (c) 2025 STMicroelectronics.
 # All rights reserved.
 #
 # This software is licensed under SLA0044 terms that can be found here:
 # https://www.st.com/resource/en/license_agreement/SLA0044.txt
 #
 # imx335 configuration file for the STM32 dcmipp IPA.
-#
-# THIS FILE WAS GENERATED BY THE STM32 ISP IQTune ON 2024-10-09 12:13:15
+# Tuned with the JudgeII Light Box
 #
 
 %YAML 1.1
@@ -22,18 +21,18 @@ algorithms:
       SensorDelay: 3
   - Awb:
       FixedGainEnable: true
-      FixedGainR: 220000000
+      FixedGainR: 182000000
       FixedGainG: 100000000
-      FixedGainB: 180000000
+      FixedGainB: 212000000
       FixedCconvEnable: true
-      FixedCconv: [180080000, -64840000, -15230000, -35550000, 169920000, -34380000, 9770000, -95700000, 185940000]
+      FixedCconv: [164670000, -20970000, -43700000, -51330000, 178670000, -27339999, -12490000, -48170000, 160670000]
       AwbEnable: true
-      ProfileName: ["IMX335-A", "IMX335-TL84", "IMX335-D50", "IMX335-D65", "Free Slot"]
-      RefColorTemp: [2856, 4000, 5000, 6500, 0]
-      GainR: [140000000, 177000000, 220000000, 245000000, 0]
-      GainG: [100000000, 100000000, 100000000, 100000000, 0]
-      GainB: [275000000, 235000000, 180000000, 155000000, 0]
-      Cconv: [151460000, -102340000, 50892000, -85991000, 210980000, -24984000, 25000000, -261000000, 341000000, 155134500, -69370000, 13106000, -38671000, 167689800, -33936000, 5546200, -66769999, 159944200, 180080000, -64840000, -15230000, -35550000, 169920000, -34380000, 9770000, -95700000, 185940000, 180080000, -64840000, -15230000, -35550000, 169920000, -34380000, 9770000, -95700000, 185940000, 0, 0, 0, 0, 0, 0, 0, 0, 0]
+      ProfileName: ["JudgeII-A", "JudgeII-TL84", "JudgeII-D65", "Free slot", "Free slot"]
+      RefColorTemp: [2810, 4015, 6650, 0, 0]
+      GainR: [137000000, 182000000, 244000000, 0, 0]
+      GainG: [100000000, 100000000, 100000000, 0, 0]
+      GainB: [287000000, 212000000, 143000000, 0, 0]
+      Cconv: [159760000, -9780000, -49990000, -45530000, 171540000, -26000000, -3300000, -110120000, 213430000, 164670000, -20970000, -43700000, -51330000, 178670000, -27339999, -12490000, -48170000, 160670000, 150570000, 2440000, -53010000, -37350000, 193760000, -56420000, -11100000, -35490000, 146590000, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
   - BadPixel:
       Enable: false
       Strength: 0
diff --git a/src/ipa/dcmipp/data/imx335_judge2_light_box.yaml b/src/ipa/dcmipp/data/imx335_judge2_light_box.yaml
deleted file mode 100644
index 6e72e38c..00000000
--- a/src/ipa/dcmipp/data/imx335_judge2_light_box.yaml
+++ /dev/null
@@ -1,54 +0,0 @@
-# Copyright (c) 2025 STMicroelectronics.
-# All rights reserved.
-#
-# This software is licensed under SLA0044 terms that can be found here:
-# https://www.st.com/resource/en/license_agreement/SLA0044.txt
-#
-# imx335 configuration file for the STM32 dcmipp IPA.
-# Tuned with the Mini Light Box
-#
-
-%YAML 1.1
----
-version: 1
-algorithms:
-  - Aec:
-      AeEnable: true
-      AntiFlickerFreq: 0
-      ExposureValue: 0.0
-      SensorDelay: 3
-  - Awb:
-      FixedGainEnable: true
-      FixedGainR: 1.77
-      FixedGainG: 1.0
-      FixedGainB: 2.35
-      FixedCconvEnable: true
-      FixedCconv: [155134500, -69370000, 13106000, -38671000, 167689800, -33936000, 5546200, -66769999, 159944200]
-      AwbEnable: true
-      ProfileName: ["JudegeII-A", "JudegeII-TL84", "JudgeII-D65", "Free slot", "Free slot"]
-      RefColorTemp: [2810, 4015, 6650, 0, 0]
-      GainR: [137000000, 182000000, 244000000, 0, 0]
-      GainG: [100000000, 100000000, 100000000, 0, 0]
-      GainB: [287000000, 212000000, 143000000, 0, 0]
-      Cconv: [159760000, -9780000, -49990000, -45530000, 171540000, -26000000, -3300000, -110120000, 213430000, 164670000, -20970000, -43700000, -51330000, 178670000, -27339999, -12490000, -48170000, 160670000, 150570000, 2440000, -53010000, -37350000, 193760000, -56420000, -11100000, -35490000, 146590000, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
-  - BadPixel:
-      Enable: false
-      Strength: 0
-      Threhsold: 0
-  - BlackLevelCorrection:
-      Enable: true
-      Level: [12, 12, 12]
-  - Contrast:
-      Enable: false
-      LuminanceFactor: [100, 100, 100, 100, 100, 100, 100, 100, 100]
-  - Demosaicing:
-      Enable: true
-      Filter: [2, 4, 4, 6]
-  - Statistic:
-      Area: [648, 486, 1296, 972]
-      Profile: 2
-  - GammaCorrection:
-      Enable: true
-# NOTE : parameters of 'statRemoval' are not handled in libcamera
-# NOTE : parameters of 'decimation' are not handled in libcamera
-# NOTE : parameter 'demosaicing.type' is not handled in libcamera
diff --git a/src/ipa/dcmipp/data/imx335_mini_light_box.yaml b/src/ipa/dcmipp/data/imx335_mini_light_box.yaml
index 11f1949d..3303cab9 100644
--- a/src/ipa/dcmipp/data/imx335_mini_light_box.yaml
+++ b/src/ipa/dcmipp/data/imx335_mini_light_box.yaml
@@ -19,9 +19,9 @@ algorithms:
       SensorDelay: 3
   - Awb:
       FixedGainEnable: true
-      FixedGainR: 1.77
-      FixedGainG: 1.0
-      FixedGainB: 2.35
+      FixedGainR: 177000000
+      FixedGainG: 100000000
+      FixedGainB: 235000000
       FixedCconvEnable: true
       FixedCconv: [155134500, -69370000, 13106000, -38671000, 167689800, -33936000, 5546200, -66769999, 159944200]
       AwbEnable: true
diff --git a/src/ipa/dcmipp/data/meson.build b/src/ipa/dcmipp/data/meson.build
index 2a4f455c..aa984efc 100644
--- a/src/ipa/dcmipp/data/meson.build
+++ b/src/ipa/dcmipp/data/meson.build
@@ -1,7 +1,9 @@
 # SPDX-License-Identifier: CC0-1.0
 
 conf_files = files([
+    'uncalibrated.yaml',
     'imx335.yaml',
+    'vd56g3.yaml',
 ])
 
 install_data(conf_files,
diff --git a/src/ipa/dcmipp/data/uncalibrated.yaml b/src/ipa/dcmipp/data/uncalibrated.yaml
new file mode 100644
index 00000000..d15cbc6f
--- /dev/null
+++ b/src/ipa/dcmipp/data/uncalibrated.yaml
@@ -0,0 +1,21 @@
+# Copyright (c) 2025 STMicroelectronics.
+# All rights reserved.
+#
+# This software is licensed under SLA0044 terms that can be found here:
+# https://www.st.com/resource/en/license_agreement/SLA0044.txt
+#
+# Empty configuration file for the STM32 dcmipp IPA.
+#
+
+%YAML 1.1
+---
+version: 1
+algorithms:
+  - Aec:
+  - Awb:
+  - BadPixel:
+  - BlackLevelCorrection:
+  - Contrast:
+  - Demosaicing:
+  - Statistic:
+  - GammaCorrection:
diff --git a/src/ipa/dcmipp/data/vd56g3.yaml b/src/ipa/dcmipp/data/vd56g3.yaml
new file mode 100644
index 00000000..62c43323
--- /dev/null
+++ b/src/ipa/dcmipp/data/vd56g3.yaml
@@ -0,0 +1,54 @@
+# Copyright (c) 2024 STMicroelectronics.
+# All rights reserved.
+#
+# This software is licensed under SLA0044 terms that can be found here:
+# https://www.st.com/resource/en/license_agreement/SLA0044.txt
+#
+# vd56g3 configuration file for the STM32 dcmipp IPA.
+#
+# THIS FILE WAS GENERATED BY THE STM32 ISP IQTune ON 2025-05-19 09:42:55
+#
+
+%YAML 1.1
+---
+version: 1
+algorithms:
+  - Aec:
+      AnalogueGain_dB: 0.0
+      ExposureTime: 9308
+      AeEnable: true
+      ExposureValue: 0.0
+      AntiFlickerFreq: 0
+      SensorDelay: 4
+  - Awb:
+      FixedGainEnable: true
+      FixedGainR: 156000000
+      FixedGainG: 100000000
+      FixedGainB: 150000000
+      FixedCconvEnable: true
+      FixedCconv: [146010000, -39280000, -14060000, -26750000, 152490000, -42520000, 1160000, -55410000, 143910000]
+      AwbEnable: true
+      ProfileName: ["JudgeII-A", "JudgeII-TL84", "JudgeII-DAY", "Free Slot", "Free Slot"]
+      RefColorTemp: [2750, 4150, 6750, 0, 0]
+      GainR: [95000000, 117000000, 156000000, 0, 0]
+      GainG: [100000000, 100000000, 100000000, 0, 0]
+      GainB: [238000000, 189000000, 150000000, 0, 0]
+      Cconv: [133939999, -20660000, -31280000, -37890000, 149680000, -26179999, 2040000, -89240000, 221830000, 147680000, -38330000, -29360000, -40320000, 146010000, -31400000, 1100000, -61240000, 174790000, 146010000, -39280000, -14060000, -26750000, 152490000, -42520000, 1160000, -55410000, 143910000, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
+  - BadPixel:
+      Enable: false
+      Strength: 0
+      Threhsold: 0
+  - BlackLevelCorrection:
+      Enable: true
+      Level: [16, 16, 16]
+  - Contrast:
+      Enable: false
+      LuminanceFactor: [0, 0, 0, 0, 0, 0, 0, 0, 0]
+  - Demosaicing:
+      Enable: true
+      Filter: [2, 4, 4, 6]
+  - Statistic:
+      Area: [140, 341, 840, 682]
+      Profile: 2
+  - GammaCorrection:
+      Enable: true
diff --git a/src/ipa/dcmipp/dcmipp.cpp b/src/ipa/dcmipp/dcmipp.cpp
index 9aa30cd1..eea008f7 100644
--- a/src/ipa/dcmipp/dcmipp.cpp
+++ b/src/ipa/dcmipp/dcmipp.cpp
@@ -112,7 +112,8 @@ int IPADcmipp::init(const IPASettings &settings, unsigned int hwRevision,
 	sensorControls_ = sensorControls;
 	camHelper_ = CameraSensorHelperFactoryBase::create(settings.sensorModel);
 	if (!camHelper_)
-		LOG(IPADcmipp, Warning) << "No camera sensor helper for " << settings.sensorModel;
+		LOG(IPADcmipp, Error) << "No camera sensor helper for " << settings.sensorModel
+				      << ". AEC will be disabled";
 
 	/* Get sensor exposure/time properties */
 	setSensorProperties(sensorInfo, sensorControls);
@@ -150,6 +151,11 @@ int IPADcmipp::start()
 	ControlList v4l2SensorControls(sensorControls_), sensorControls;
 	sensorControls.set(controls::draft::AnalogueGain_dB, context_.sensor.gain);
 	sensorControls.set(controls::ExposureTime, context_.sensor.exposure);
+
+	/* If needed, disable sensor Auto Exposure */
+	if (context_.info.doDisableSensorAE)
+		v4l2SensorControls.set(V4L2_CID_EXPOSURE_AUTO, (int32_t)V4L2_EXPOSURE_MANUAL);
+
 	convertControls(sensorControls, v4l2SensorControls);
 	setSensorControls.emit(0, v4l2SensorControls);
 
@@ -186,6 +192,9 @@ int IPADcmipp::configure(const IPACameraSensorInfo &sensorInfo,
 	/* Update the camera controls using the new sensor settings. */
 	updateControls(ipaControls);
 
+	/* Define whether the sensor auto exposure (if supported) shall be disabled */
+	context_.info.doDisableSensorAE = context_.info.AECEnable && sensorControls.find(V4L2_CID_EXPOSURE_AUTO) != sensorControls.end();
+
 	return 0;
 }
 
@@ -329,16 +338,18 @@ float IPADcmipp::dBToLinear(float gain)
 
 void IPADcmipp::convertControls(ControlList &controls, ControlList &v4l2Controls)
 {
-	/* Converts exposure to V4L2 control unit (number of lines) */
-	const auto &exposure = controls.get(controls::ExposureTime);
-	if (exposure)
-		v4l2Controls.set(V4L2_CID_EXPOSURE, (int32_t)(*exposure / context_.info.sensorLineDuration_us));
-
-	/* Converts gain to V4L2 control unit (eg 0.3 dB step for IMX335 sensor) */
-	const auto &gain = controls.get(controls::draft::AnalogueGain_dB);
-	if (gain && camHelper_) {
-		float gainLinear = dBToLinear(*gain);
-		v4l2Controls.set(V4L2_CID_ANALOGUE_GAIN, (int32_t)(camHelper_->gainCode(gainLinear)));
+	if (camHelper_) {
+		/* Converts exposure to V4L2 control unit (number of lines) */
+		const auto &exposure = controls.get(controls::ExposureTime);
+		if (exposure)
+			v4l2Controls.set(V4L2_CID_EXPOSURE, (int32_t)(*exposure / context_.info.sensorLineDuration_us));
+
+		/* Converts gain to V4L2 control unit (eg 0.3 dB step for IMX335 sensor) */
+		const auto &gain = controls.get(controls::draft::AnalogueGain_dB);
+		if (gain) {
+			float gainLinear = dBToLinear(*gain);
+			v4l2Controls.set(V4L2_CID_ANALOGUE_GAIN, (int32_t)(camHelper_->gainCode(gainLinear)));
+		}
 	}
 
 	/* Converts statistic area */
diff --git a/src/ipa/dcmipp/ipa_context.h b/src/ipa/dcmipp/ipa_context.h
index 3a826f1c..dacc38e4 100644
--- a/src/ipa/dcmipp/ipa_context.h
+++ b/src/ipa/dcmipp/ipa_context.h
@@ -117,6 +117,7 @@ struct IPAContext {
 		int64_t frameDurationsMin; /* Min frame duration in microseconds */
 		int64_t frameDurationsMax; /* Max frame duration in microseconds */
 		int64_t frameDurationsDef; /* Default frame duration in microseconds */
+		bool doDisableSensorAE; /* Whether the sensor AE shall be disabled or not */
 	} info;
 
 	/* Supported controls */
diff --git a/src/ipa/libipa/camera_sensor_helper.cpp b/src/ipa/libipa/camera_sensor_helper.cpp
index 2cd61fcc..6785e272 100644
--- a/src/ipa/libipa/camera_sensor_helper.cpp
+++ b/src/ipa/libipa/camera_sensor_helper.cpp
@@ -631,6 +631,17 @@ public:
 };
 REGISTER_CAMERA_SENSOR_HELPER("ov13858", CameraSensorHelperOv13858)
 
+class CameraSensorHelperVd56g3 : public CameraSensorHelper
+{
+public:
+	CameraSensorHelperVd56g3()
+	{
+		gainType_ = AnalogueGainLinear;
+		gainConstants_.linear = { 0, 32, -1, 32 };
+	}
+};
+REGISTER_CAMERA_SENSOR_HELPER("vd56g3", CameraSensorHelperVd56g3)
+
 #endif /* __DOXYGEN__ */
 
 } /* namespace ipa */
diff --git a/src/libcamera/pipeline/dcmipp/dcmipp.cpp b/src/libcamera/pipeline/dcmipp/dcmipp.cpp
index a7353fe2..a8fadee8 100644
--- a/src/libcamera/pipeline/dcmipp/dcmipp.cpp
+++ b/src/libcamera/pipeline/dcmipp/dcmipp.cpp
@@ -552,14 +552,18 @@ PipelineHandlerDcmipp::generateConfiguration(Camera *camera,
 	 * Raw: targetting RawBayer extraction of the image from a sensor. Format depends on
 	 *      the format of the sensor. This is only done by the Dump pipe.
 	 */
+	Size sizeSensor = data->sensor_->resolution();
+	Size maxSizeVideoRecording(1920, 1080);
+	Size maxSizeViewfinder(1024, 600);
 	for (const StreamRole role : roles) {
 		DcmippPath *path;
 		Size maxSize;
 		PixelFormat pixelFormat;
 
+
 		switch (role) {
 		case StreamRole::StillCapture:
-			maxSize = data->sensor_->resolution();
+			maxSize = sizeSensor;
 			pixelFormat = formats::YUV420;
 			if (mainAllocated) {
 				LOG(DCMIPP, Error) << "Main pipe already allocated, ABORT";
@@ -569,7 +573,7 @@ PipelineHandlerDcmipp::generateConfiguration(Camera *camera,
 			mainAllocated = true;
 			break;
 		case StreamRole::Viewfinder:
-			maxSize = { 1024, 600 };
+			maxSize = maxSizeViewfinder.boundedTo(sizeSensor);
 			pixelFormat = formats::RGB888;
 			if (auxAllocated) {
 				LOG(DCMIPP, Error) << "Aux pipe already allocated, ABORT";
@@ -579,7 +583,7 @@ PipelineHandlerDcmipp::generateConfiguration(Camera *camera,
 			auxAllocated = true;
 			break;
 		case StreamRole::VideoRecording:
-			maxSize = { 1920, 1080 };
+			maxSize = maxSizeVideoRecording.boundedTo(sizeSensor);
 			pixelFormat = formats::NV12;
 			if (mainAllocated) {
 				LOG(DCMIPP, Error) << "Main pipe already allocated, ABORT";
@@ -589,7 +593,7 @@ PipelineHandlerDcmipp::generateConfiguration(Camera *camera,
 			mainAllocated = true;
 			break;
 		case StreamRole::Raw:
-			maxSize = data->sensor_->resolution();
+			maxSize = sizeSensor;
 			/*
 			 * In case of Raw role, the pixelFormat is actually given by the
 			 * dcmipp_path generateConfiguration since it depends on sensor format
@@ -1086,6 +1090,8 @@ bool PipelineHandlerDcmipp::match(DeviceEnumerator *enumerator)
 		ipaTuningFile = std::string(configFromEnv);
 	} else {
 		ipaTuningFile = data->ipa_->configurationFile(data->sensor_->model() + ".yaml");
+		if (ipaTuningFile.empty())
+			ipaTuningFile = data->ipa_->configurationFile("uncalibrated.yaml");
 	}
 
 	if (data->ipa_->init(IPASettings{ ipaTuningFile, data->sensor_->model() },
diff --git a/src/libcamera/sensor/camera_sensor_properties.cpp b/src/libcamera/sensor/camera_sensor_properties.cpp
index b18524d8..1b2a29ae 100644
--- a/src/libcamera/sensor/camera_sensor_properties.cpp
+++ b/src/libcamera/sensor/camera_sensor_properties.cpp
@@ -267,6 +267,10 @@ const CameraSensorProperties *CameraSensorProperties::get(const std::string &sen
 				{ controls::draft::TestPatternModeColorBars, 1 },
 			},
 		} },
+		{ "vd56g3", {
+			.unitCellSize = { 2610, 2610 },
+			.testPatternModes = {},
+		} },
 	};
 
 	const auto it = sensorProps.find(sensor);
-- 
2.25.1

